akka-actor/src/main/scala/akka/dispatch/Mailbox.scala

```diff
@@ -207,10 +207,11 @@ private[akka] abstract class Mailbox(val messageQueue: MessageQueue)
     new LatestFirstSystemMessageList(Unsafe.instance.getObjectVolatile(this, AbstractMailbox.systemMessageOffset).asInstanceOf[SystemMessage])
 
   protected final def systemQueuePut(_old: LatestFirstSystemMessageList, _new: LatestFirstSystemMessageList): Boolean =
-    // Note: calling .head is not actually existing on the bytecode level as the parameters _old and _new
-    // are SystemMessage instances hidden during compile time behind the SystemMessageList value class.
-    // Without calling .head the parameters would be boxed in SystemMessageList wrapper.
-    Unsafe.instance.compareAndSwapObject(this, AbstractMailbox.systemMessageOffset, _old.head, _new.head)
+    _old.head == _new.head ||
```

Comment:
equals on system messages seems heavy, and is it even correct? should it be `eq`?

Reply:
Oh, yes, stupid. Usually, both are `null` but it should `eq`, definitely.

```suggestion
    _old.head eq _new.head ||
```